src/ciflow-bot.ts

```diff
@@ -0,0 +1,220 @@
+
+    // Logging of the dispatch
+    this.ctx.log.info(
+      {
+        dispatch_labels: this.dispatch_labels,
+        dispatch_strategies: this.dispatch_strategies,
+        event: this.event,
+        owner: this.owner,
+        pr_number: this.pr_number,
+        pr_labels: this.pr_labels,
+        repo: this.repo
+      },
+      'ciflow dispatch success!'
+    );
+  }
+
+  dispatch_strategy_func(strategyName: string): void {
+    switch (strategyName) {
+      case 'add_default_labels':
+        // Default dispatch algorithm: 'add_default_labels'
+        // Just make sure the we add a 'ciflow/default' to the existing set of pr_labels
+        if (this.dispatch_labels.length === 0) {
+          this.dispatch_labels = this.pr_labels;
+        }
+        this.dispatch_labels = ['ciflow/default', ...this.dispatch_labels];
+        break;
+      default: {
+        this.ctx.log.error({strategyName}, 'Unknown dispatch strategy');
+        break;
+      }
+    }
+  }
+
+  // signal_github sends a signal to GitHub to trigger the dispatch
+  // The logic here is leverage some event that's rarely triggered by other users or bots,
+  // thus we pick "assign/unassign" to begin with. See details from the CIFlow RFC:
+  // https://github.com/pytorch/pytorch/issues/61888
+  async signal_github(): Promise<void> {
+    await this.setLabels();
+
+    await this.ctx.github.issues.addAssignees({
+      owner: this.owner,
+      repo: this.repo,
+      issue_number: this.pr_number,
+      assignees: [this.bot_assignee]
+    });
+
+    await this.ctx.github.issues.removeAssignees({
+      owner: this.owner,
+      repo: this.repo,
+      issue_number: this.pr_number,
+      assignees: [this.bot_assignee]
+    });
+  }
+
+  async setLabels(): Promise<void> {
+    const labels = this.dispatch_labels.filter(label =>
+      label.startsWith(this.pr_label_prefix)
+    );
+    const labelsToDelete = this.pr_labels.filter(l => !labels.includes(l));
+    const labelsToAdd = labels.filter(l => !this.pr_labels.includes(l));
+    for (const label of labelsToDelete) {
+      await this.ctx.github.issues.removeLabel({
+        owner: this.ctx.payload.repository.owner.login,
+        repo: this.ctx.payload.repository.name,
+        issue_number: this.pr_number,
+        name: label
+      });
+    }
+    await this.ctx.github.issues.addLabels({
+      owner: this.ctx.payload.repository.owner.login,
+      repo: this.ctx.payload.repository.name,
+      issue_number: this.pr_number,
+      labels: labelsToAdd
+    });
+    this.dispatch_labels = labels;
+  }
+
+  parseComment(): void {
+    const re = new RegExp(`^.*@${this.bot_app_name}\\s+(\\w+)\\s?(.*)$`);
+    const found = this.comment_body?.match(re);
+    if (!found) {
+      return;
+    }
+
+    if (found.length >= 2) {
+      this.command = found[1];
+    }
+    if (found.length === 3) {
+      this.command_args = found[2].split(' ');
+    }
+  }
+
+  parseContext(ctx: probot.Context): void {
```

Comment:
Wouldn't it be better if this function returned something like `ParseContext` class rather than mutated self? (Considering that it is called from async method, which means several instances of the method can execute concurrently, aren't they?)

Reply:
refactor the code a little bit to always construct a new instance for new webhook event - thus `setContext()` only called once within the same event loop of the new instance.

mutating self helps the dispatch algorithms to gather the information needed within the class.